geoips/plugins/modules/readers/generic_satpy.py

```diff
@@ -0,0 +1,145 @@
+# # # This source code is protected under the license referenced at
+# # # https://github.com/NRLMMD-GEOIPS.
+
+
+"""Generic Satpy reader.
+
+Provides a GeoIPS wrapper for Satpy reader functionality. For more detailed
+documentation of Satpy, see the documentation here:
+https://satpy.readthedocs.io/en/stable/. This reader creates a Scene object with
+the provided filenames and specified reader, then loads provided channels. The
+satpy_reader, and fnames arguments are passed to the Scene constructor.  The
+Scene object is then reorganized into a Dictionary of XArray Dataset objects as
+required by the GeoIPS reader interface.  The various additional arguments to
+the Scene constructor and its load method are provided as optional arguments to
+the call function in this plugin.  This allows any arguments required by a
+particular Satpy reader to be passed via the GeoIPS command line interface
+reader_kwargs argument. The reader_kwargs argument should provide a string
+containing a JSON Object mapping call arguments with their desired values.
+
+In order to map the structure of a Satpy Scene to the Dictionary of Datasets, an
+additional argument named "channel_groups" should be provided via reader_kwargs.
+The argument is a dictionary that groups desired channels by resolution.  This
+dictionary's keys should be strings representing the name of each group and the
+values should be a list of channel names that have the same resolution. The name
+of each group will be used as the keys of the output dictionary and the list of
+channels will be grouped together as a Dataset stored as the associated value to
+the key.  So for example the following channel_groups JSON Object could be used
+to read a selection of visible and IR GOES16 data:
+"channel_groups":{"0.5km":["C02"], "2km":["C13", "C14"]}
+
+While Satpy provides a very large number of readers, a very small number of them
+have been used with this plugin.  This plugin should be considered experimental
+and is not guaranteed to work with all of Satpy's readers.
+"""
+
+import logging
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import dask
+import pyresample.geometry
+import satpy
+import xarray as xr
+
+log = logging.getLogger(__name__)
+
+interface = "readers"
+family = "standard"
+name = "generic_satpy"
+
+
+DEFAULT_COORD_CHUNK = 4096
+
+
+def call(
+    fnames: List[str],
+    source_name: str,
+    satpy_reader: str,
+    channel_groups: Dict[str, List[str]],
+    platform_name: Optional[str] = None,
```

Comment:
Why is `platform_name` needed? Is GeoIPS really strict on what these can be? Satpy's returned DataArrays should have a platform name in the `.attrs` metadata.

Reply:
I think at the time I started this, it was required by GeoIPS, but it doesn't look like it is now.  If @mindyls or @jsolbrig can confirm that the `platform_name` isn't a required attribute in the output Dataset then I think it makes sense to remove it.

Reply:
It looks like the unit tests require the `platform_name` to be present in each output `Dataset`.

Reply:
platform_name is a required attribute on our xarray dataset outputs from the readers: https://github.com/NRLMMD-GEOIPS/geoips/blob/main/docs/source/devguide/xarray_standards.rst 

It is not strictly enforced what these values are, like it is for source_names, though having consistent platform names throughout GeoIPS would probably good...  So we'd probably want a mapping somewhere of satpy platform names to geoips platform names.

Reply:
Right now the plugin will look up the `platform_name` from the attributes of the first channel it encounters.  By default, it will use the `platform_name` key but this can be changed using the command line arguments. Additionally, you can optionally just specify the platform name via the command line arguments if necessary.